File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class SlidingWindowMaximum {
    int limit;
    deque<pair<int, int>> list;
    public:
    SlidingWindowMaximum(int k) {
        limit = k;
    }

    void insert(int index, int value) {
        while (!list.empty() && list.back().first < value) list.pop_back();
        list.push_back({value, index});
    }

    int get(int index) {
        int boundry = index-limit+1;
        while (list.front().second < boundry) list.pop_front();
        return list.front().first;
    }
};

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        SlidingWindowMaximum ds(k);
        vector<int> vec;

        for (int i=0; i<k; i++) ds.insert(i, nums[i]);
        vec.push_back(ds.get(k-1));

        for(int i=k; i<nums.size(); i++) {
            ds.insert(i, nums[i]);
            vec.push_back(ds.get(i));
        }

        return vec;
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.cs
public class SlidingWindowMaximum {
    private int limit;
    private readonly LinkedList<(int, int)> list;

    public SlidingWindowMaximum(int limit) {
        this.limit = limit;
        list = new LinkedList<(int, int)>();
    }

    public void Insert(int index, int value) {
        while (list.Count > 0 && list.Last.Value.Item1 < value) {
            list.RemoveLast();
        }

        list.AddLast((value, index));
    }

    public int Get(int index) {
        var boundry = index-limit+1;
        while (list.First.Value.Item2 < boundry) list.RemoveFirst();
        return list.First.Value.Item1;
    }
}

public class Solution {
    public int[] MaxSlidingWindow(int[] nums, int k) {
        var maximum = new SlidingWindowMaximum(k);
        int[] result = new int[nums.Count()-k+1];

        for (int i=0; i<k; i++) {
            maximum.Insert(i, nums[i]);
        }
        result[0] = maximum.Get(k-1);

        for (int i=k; i<nums.Count(); i++) {
            maximum.Insert(i, nums[i]);
            result[i-k+1] = maximum.Get(i);
        }

        return result;
    }
}
